### ISSUE TITLE
Array of Record Types fails to type check

### ISSUE THREAD

**Issue by timzwiebel** [google/closure-compiler#2570]
The old type checking doesn't work correctly with an array of record types. I would expect to get an error with --jscomp_error=checkTypes similar to the one that I do with --jscomp_error=newCheckTypes.

For example:
```
$ cat foo.js
/** @param {function(!Array<!{a:string}>):void} g */
function f(g) {}
/** @param {!Array<!{a:string,b:number}>} x */
function g(x) {}
// The following should produce a type error since the record type in g's
// parameter isn't a subset of the record type in f's paramter.
f(g);
```
```
$ java -jar closure-compiler-v20170626.jar --js foo.js --jscomp_error=checkTypes
function f(a){}function g(a){}f(g);
```
```
$ java -jar closure-compiler-v20170626.jar --js foo.js --jscomp_error=newCheckTypes
foo.js:7: ERROR - Invalid type for parameter 1 of function f.
Expected : function(Array<{a: string}>):undefined
Found    : function(Array<{a: string, b: number}>):undefined
More details:
The expected and found types are functions which have incompatible types for argument 1.
Expected a supertype of : Array<{a: string}>
but found               : Array<{a: string, b: number}>
f(g);
  ^

1 error(s), 0 warning(s)
```

**Comment by timzwiebel** [comment]
A similar problem exists when `f` uses `!Object` and `g` uses `!{a:string}`. I would expect a type error with the old type checking, but get none.
```
$ cat foo.js
/** @param {function(!Array<!Object>):void} g */
function f(g) {}
/** @param {!Array<!{a:string}>} x */
function g(x) {}
// The following should produce a type error since the record type in g's
// parameter isn't a subset of the record type in f's paramter.
f(g);
```

**Comment by timzwiebel** [comment]
And just to be clear, both of the previous examples work as expected when not using Arrays.

**Comment by dimvar** [comment]
Array generics are loose in the old type checker. This is a known issue, unlikely to fix. People will get the correct behavior as they migrate to the new type checker.
